lib/puppet/network/http/connection.rb

```diff
@@ -189,6 +189,32 @@ def request_with_redirects(request, options)
             request.each do |header, value|
               current_request[header] = value
             end
+          when 503, 429
+            retry_after = current_response['Retry-After']
+
+            if retry_after.nil?
+              # Set return value and advance to next loop.
+              response = current_response
+              next
+            end
+
+            begin
+              retry_sleep = Integer(retry_after)
```

Comment:
In that case, the Puppet agent would fail and move onto the next request and eventually onto the next Puppet Run. I would actually recommend against this sort of load balancer configuration if the Retry-After times weren't randomized -- as that would cause a thundering herd to form.

Puppet Server includes a randomized Retry-After with its 503 responses which should have the effect of breaking agent herds up.